ARROW-10667: [Rust] [Parquet] Add a convenience type for writing Parquet to memory

carols10cents opened the pull request:
Similar to the [`SliceableCursor`](https://github.com/apache/arrow/blob/0e841aa666b637e24e2889acab7621aa01fb7bcf/rust/parquet/src/util/cursor.rs#L22-L27) type that provides a convenience for reading Parquet from memory, I would like to propose a type to make it convenient to write Parquet to memory.

This is possible for clients to implement today, but seems common enough to want to provide for everyone to use.

alamb commented on rust/parquet/src/util/cursor.rs:
```diff
@@ -129,6 +132,56 @@ impl Seek for SliceableCursor {
+impl TryClone for WriteableCursor {
+    fn try_clone(&self) -> std::io::Result<Self> {
+        Ok(Self {
+            buffer: self.buffer.clone(),
```
THis is clever. 👍

alamb reviewed (approved):
I think this would be super helpful (and we plan to use this code in IOx downstream). 

@nevi-me  @rdettai  or @sunchao what do you think?

sunchao commented on rust/parquet/src/util/cursor.rs:
```diff
@@ -129,6 +132,56 @@ impl Seek for SliceableCursor {
 
+/// Use this type to write Parquet to memory rather than a file.
+#[derive(Debug, Default, Clone)]
+pub struct WriteableCursor {
```
nit: it's not easy to see this writes to memory - maybe we can make it more explicit in naming like `InMemoryWriteableCursor` or similar?

sunchao reviewed (approved):
LGTM just one nit. Thanks @carols10cents

carols10cents commented:
@sunchao Thank you for the suggestion for renaming the type to `InMemoryWriteableCursor`! I do think that's clearer. I've renamed the type, rebased this branch, and resolved merge conflicts, so I think this is ready to go pending CI.
